### PR TITLE
moveit_servo: Only halt the joints that violate the position limits

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

Previously when one joint is close to a limit moveit_servo will halt all the joints in the group, this PR make it so it only halts the joints that are close to the limit

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h [hunk 1]
```diff
@@ -151,13 +151,16 @@ class ServoCalcs
...
+  void suddenHalt(sensor_msgs::msg::JointState& joint_state,
+                  const std::vector<const moveit::core::JointModel*>& joints_to_halt) const;
 
   /** \brief  Scale the delta theta to match joint velocity/acceleration limits */
   void enforceVelLimits(Eigen::ArrayXd& delta_theta);
 
-  /** \brief Avoid overshooting joint limits */
-  bool enforcePositionLimits(sensor_msgs::msg::JointState& joint_state) const;
+  /** \brief Avoid overshooting joint limits
+      \return Vector of the joints that violate the position limits
```
Comment (JafarAbdi) [inline moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:161]: Should I apply the following suggestion .?

```suggestion
      \return Vector of the joints that are close to the position limits
```

Comment (schornakj) [reply moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:161]: I think your revised wording is clearer.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp [hunk 2]
```diff
@@ -811,18 +814,23 @@ bool ServoCalcs::enforcePositionLimits(sensor_msgs::msg::JointState& joint_state
...
+  if (!joints_to_halt.empty())
+  {
+    std::ostringstream joints_names;
+    auto joiner = std::experimental::make_ostream_joiner(joints_names, ", ");
+    std::transform(joints_to_halt.cbegin(), joints_to_halt.cend(), joiner,
+                   [](const auto& joint) { return joint->getName(); });
+    rclcpp::Clock& clock = *node_->get_clock();
+    RCLCPP_WARN_STREAM_THROTTLE(LOGGER, clock, ROS_LOG_THROTTLE_PERIOD,
+                                node_->get_name()
+                                    << " " << joints_names.str() << " close to a position limit. Halting.");
```
Comment (schornakj) [inline moveit_ros/moveit_servo/src/servo_calcs.cpp:831]: I usually write this like below. Is there a specific reason you did it the other way?

```suggestion
    RCLCPP_WARN_STREAM_THROTTLE(LOGGER, node_->get_clock(), ROS_LOG_THROTTLE_PERIOD,
                                node_->get_name()
                                    << " " << joints_names.str() << " close to a position limit. Halting.");
```

**Review by schornakj — COMMENTED** [review]
I tried this in simulation with a UR-5e and it works like I'd expect.

**Comment by nbbrooks** [comment]
I think the current versus this new behavior should be controlled by a param. If you are servoing in Cartesian space the behavior in this PR may be highly undesirable.

File: moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h (hunk 1 above)
Comment (nbbrooks) [reply moveit_ros/moveit_servo/include/moveit_servo/servo_calcs.h:161]: My suggestion would be

```suggestion
      \return Vector of the joints that increasingly violate the position margin limits
```

The comment in the first line of the enforcePositionLimits implementation is pretty descriptive, but probably too long for /brief
> // Halt if we're past a joint margin and joint velocity is moving even farther past

File: moveit_ros/moveit_servo/src/servo_calcs.cpp [hunk 3]
```diff
@@ -811,18 +814,23 @@ bool ServoCalcs::enforcePositionLimits(sensor_msgs::msg::JointState& joint_state
...
     }
   }
-
-  return !halting;
+  if (!joints_to_halt.empty())
+  {
+    std::ostringstream joints_names;
+    auto joiner = std::experimental::make_ostream_joiner(joints_names, ", ");
+    std::transform(joints_to_halt.cbegin(), joints_to_halt.cend(), joiner,
+                   [](const auto& joint) { return joint->getName(); });
```
[the remark this answers is not in the document]
Comment (v4hn) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:827]: [MTC uses `boost::join` + `adaptors`](https://github.com/ros-planning/moveit_task_constructor/blob/master/core/src/merge.cpp#L95-L96) for the same logic. Personally I prefer boost over experimental headers, but it's a matter of taste and spelling out the concatenation is just cluttering code.

[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:827]: Personally, I would prefer using boost, but I saw that servo doesn't depend on it so I decided it's not worth it just for this line

File: moveit_ros/moveit_servo/src/servo_calcs.cpp [hunk 4]
```diff
@@ -41,6 +41,7 @@
 #include <thread>
 #include <chrono>
 #include <mutex>
+#include <experimental/iterator>
```
[the remark this answers is not in the document]
Comment (JafarAbdi) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:44]: You're right, looking at the source-code of MSVC it doesn't have this header https://github.com/microsoft/STL

**Comment by v4hn** [comment]
> If you are servoing in Cartesian space the behavior in this PR may be highly undesirable.

If the halting joints are not considered during solving for joint space motions (where the solver may theoretically choose a null-space solution that leaves the halted joint alone), this critically breaks functionality (it fails to follow the Cartesian command).
In that case would it make sense to restrict this functionality to joint servoing only?

**Comment by JafarAbdi** [comment]
> I think the current versus this new behavior should be controlled by a param. If you are servoing in Cartesian space the behavior in this PR may be highly undesirable.

Thanks for the catch, fixed!

**Comment by nbbrooks** [comment]
> > If you are servoing in Cartesian space the behavior in this PR may be highly undesirable.
> 
> If the halting joints are not considered during solving for joint space motions (where the solver may theoretically choose a null-space solution that leaves the halted joint alone), this critically breaks functionality (it fails to follow the Cartesian command).
> In that case would it make sense to restrict this functionality to joint servoing only?

I see 2 options

1. Assume that
- joint servoing -> halt only offending joints
- cart servoing -> halt all joints
(current state of PR)

2. Control which halt strategy is used via a dynamic parameter so we don't make any assumptions
- would allow for joint servoing to halt either only or all offending joints

In the case of a clever solver using null space to not use a joint near/at its limit, enforcePositionLimits would not trigger on that joint, because it requires a joint to be at/beyond its margin added limit and moving further past it.

So I think this boils down to: do we want to assume the behavior for joint servoing?

**Comment by AndyZe** [comment]
I like ^Nathan's option 2, where it defaults to the previous behavior

**Comment by nbbrooks** [comment]
I feel like there may be unintended side effects of changing the default behavior for people using Primary/Replica arms and think we should consider that.

If we control the behavior via parameters, we could (imo) make management easier by having two params
halt_all_joints_on_joint_mode (default true)
halt_all_joints_in_cartesian_mode (default true)
(someone can come up with better names)

Then there won't be a bunch of overhead second guessing the user's intent if they switch between joint and cart sevoing.

File: moveit_ros/moveit_servo/src/servo_calcs.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nbbrooks) [reply moveit_ros/moveit_servo/src/servo_calcs.cpp:837]: What about ostream_iterator?

**Review by nbbrooks — APPROVED** [review]
I tested this with an xbox controller and the moveit2_tutorial, where I put the panda link7 into a limit and then tried moving link6 or doing Cartesian movement. 

`halt_all_joints_in_joint_mode: false` works as expected
`halt_all_joints_in_cartesian_mode: false` is more for future proofing but I was curious what the behavior would look like. It is somewhat hard to cause this to trigger, as (at least with default singularity thresholds), the arm is stopped by singularity thresholds before it would be stopped by joint limits. But you can trigger it in certain cases and the Cartesian movement continues, but just doesn't track as well in Cartesian space (as expected).

**Pull request merged**